[AArch64] move extension information into tablgen

tommat01 opened the pull request:
Generate target features and FMVExtensions from tablegen.

Use MArchName/ArchKindEnumSpelling to avoid renamings.
Cases where there is simply a case difference are handled by
consistently uppercasing the AEK_ name in the emitted code.

Remove some Extensions which were not needed.
These had AEK entries but were never actually used for anything.
They are not present in Extensions[] data.

tommat01 committed:
[AArch64] move extension information into tablgen

Generate target features and FMVExtensions from tablegen.

Use MArchName/ArchKindEnumSpelling to avoid renamings.
Cases where there is simply a case difference are handled by
consistently uppercasing the AEK_ name in the emitted code.

Remove some Extensions which were not needed.
These had AEK entries but were never actually used for anything.
They are not present in Extensions[] data.

tommat01 committed:
Override FMVDependencies where necessary

tommat01 committed:
clang-format

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen

jthackray commented on llvm/lib/Target/AArch64/AArch64Features.td:
```diff
@@ -11,42 +11,123 @@
+    string FMVBit = _FMVBit;
+
+    // List of features that this feature depends on.
+    // FIXME generate this from Implies.
```
Presumably this FIXME is coming in a future PR?

tmatheson-arm replied:
That would require the FMV dependencies being aligned with the existing `SubtargetFeature` dependencies, which I believe is being considered but is a WIP.

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen

tommat01 committed:
clarify that memtag means different things for FMV and -march

tommat01 committed:
remove BTI Note

tmatheson-arm commented on llvm/lib/Target/AArch64/AArch64Features.td:
```diff
@@ -426,30 +532,40 @@ def FeatureSpecRestrict : SubtargetFeature<"specrestrict", "HasSpecRestrict",
 def FeatureMTE : Extension<"mte", "MTE",
-    "Enable Memory Tagging Extension (FEAT_MTE, FEAT_MTE2)" >;
+    "Enable Memory Tagging Extension (FEAT_MTE, FEAT_MTE2)", [],
+    "FEAT_MEMTAG", "", 440>;
```
Good point. There should be no behaviour change here caused by this patch. I've added a comment in f1d287a75139bd8a09ee9d596b10aa77a2b672b7 explaining that `"memtag"` means different things for FMV and non-FMV extensions. I've left the `SubtargetFeature` description as-is because I think it is most likely to be used for printing command line help in the future, and the FMV difference will not be relevant there.

tommat01 committed:
Fix the ordering of SME/SVE target features in tests

These change because the order of the definitions of the extensions
has changes since they are autogenerated.

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen
